[Jetpack Focus] Revamp Landing Screen - Sensor animation

ovitrif opened the pull request:
Fixes #17075

With this PR we're using accelerometer sensor data (y-axis acceleration) to influence the speed of the vertical scrolling text from the background of the Jetpack App Landing Screen.

~~⚠️ **Important Notice** This PR should target `trunk` after `feature/revamp-landing-screen` is merged.
PR #17203 should be merged first, then we can target `trunk` with this PR and set it `Ready for review`.~~

To test:

<details>
<summary><b>Prerequisite</b> - Enable the <code>LandingScreenRevamp</code> feature flag</summary>

- Either `login` → `Me` → `App Settings` → `Debug Settings` → enable `LandingScreenRevamp` → `RESTART THE APP` → `logout` or
- Apply this hard-coded solution:

<details>
<summary>Patch to enable the <code>LANDING_SCREEN_REVAMP</code> build flag</summary>

```diff
diff --git a/WordPress/build.gradle b/WordPress/build.gradle
index 63260a4ca8..fe999fa92f 100644
--- a/WordPress/build.gradle
+++ b/WordPress/build.gradle
@@ -112,7 +112,7 @@ android {
         buildConfigField "boolean", "READER_COMMENTS_MODERATION", "false"
         buildConfigField "boolean", "SITE_INTENT_QUESTION", "true"
         buildConfigField "boolean", "SITE_NAME", "false"
-        buildConfigField "boolean", "LANDING_SCREEN_REVAMP", "false"
+        buildConfigField "boolean", "LANDING_SCREEN_REVAMP", "true"
         buildConfigField "boolean", "LAND_ON_THE_EDITOR", "false"
         buildConfigField "boolean", "BLOGGING_PROMPTS", "false"
         buildConfigField "boolean", "QUICK_START_EXISTING_USERS_V2", "false"
```  
</details>

---

</details>

##### 1️⃣ Test Case 1: Happy Flow

0. Open the JP app and make sure you're logged out
1. Make sure you're holding the device in your hand slightly rotated vertically, as illustrated:
   <img width="100" src="https://user-images.githubusercontent.com/4588074/191999204-611285aa-9bc0-4cdc-8384-bed2ef6f07ff.png">
3. **Expect** the large text on the new login screen to scroll down by default at an accelerated speed
4. Lay the device flat on your office table
5. **Expect** The text to scroll **down** at a slower speed
6. Pick up the device and rotate it slightly inverse than in the illustration (top of the phone should be below its bottom)
7. **Expect** The text to scroll **up**

##### 2️⃣ Test Case 2: Feature flag off

0. Disable the <code>LandingScreenRevamp</code> feature flag (revert change from **Prerequisite**)
1. Open the JP app and make sure you're logged out
2. **Expect** To be presented with the previous design
3. Tap any of the 2 buttons
4. **Expect** the Login flows to still work as usual.

##### 3️⃣ Test Case 3: Sensors off

1. Enable developer mode:
   `Settings` → `About phone / emulated device` → tap `Build number` 7 times
2. Enable `Sensors Off quick setting:
   `Go to `Settings` → `System` → `Developer options` → `Quick settings developer tiles` tap `Sensors Off`
3. Toggle Sensors off:
   Swipe down from the top of the screen and tap `Sensors Off`
4. Open the JP app and make sure you're logged out
5. **Expect** the large text to scroll down at a speed similar to the normal position speed when sensors are enabled
   (step 3 of Test Case 1)

##### 4️⃣ Test Case 4: No accelerometer sensor

1. Apply this patch to simulate missing accelerometer sensor on the device:

```diff
diff --git a/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt b/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
index 99e8570448..c57f1aaf1e 100644
--- a/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
+++ b/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
@@ -62,7 +62,7 @@ class LoginPrologueRevampedViewModel @Inject constructor(
 
         override fun onActive() {
             super.onActive()
-            val sensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
+            val sensor: Sensor? = null
             sensorManager.registerListener(this, sensor, SensorManager.SENSOR_DELAY_GAME)
         }
```

2. Open the JP app and make sure you're logged out

3. **Expect** the large text to scroll down at a speed similar to the normal position speed when sensors are enabled

<details>
<summary><h4>🎬 Video Preview</h4></summary>

https://user-images.githubusercontent.com/4588074/192004342-b2955e14-95b0-45d3-87ff-9db4f5ccce28.mp4

- Initially the device is rotated to `-45°` (reflecting a normal use position)
- At second `5` the device is rotated to `-90°` (reflecting flat on table position)
- At second `13` the device is rotated to `-105°` (inverted rotation that triggers reverse scrolling)

---

</details>

## Regression Notes
4. Potential unintended areas of impact
   New Jetpack Landing Screen when sensor is not available.

8. What I did to test those areas of impact (or what existing automated tests I relied on)
   Tested manually different configs: sensors disabled / missing and added test cases for these.

9. What automated tests I added (or what prevented me from doing so)
   None, we're interacting with platform code and UI animation code.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ovitrif committed:
Add JP app LoginPrologueRevampedFragment and hook it from LoginActivity

ovitrif committed:
Align login button to end of screen

ovitrif committed:
Add portrait resources for jetpack splash screen bg

ovitrif committed:
Add string resources for jetpack splash screen bg

ovitrif committed:
Scaffold UI of the jetpack splash screen

ovitrif committed:
Adjust logo size and padding to match design

ovitrif committed:
Add top background fading behind jetpack logo on the splash screen

ovitrif committed:
Show login prologue in full screen

ovitrif committed:
Remove dat dere additional whitespace

ovitrif committed:
Remove redundant fillMaxSize modifiers

ovitrif committed:
move method to set fullscreen inside login prologue revamped fragment

ovitrif committed:
Extract components to separate files

ovitrif committed:
Add large texts

ovitrif committed:
Fix list of jetpack features texts

ovitrif committed:
Fix auto scroll

ovitrif committed:
Rm unused AutoScrollingModel

ovitrif committed:
Remove bg_jetpack_login_splash for night mode since it's the same

ovitrif committed:
Put LoginScreenRevamped args on separate lines in preview

ovitrif committed:
Add variable colors to large texts

ovitrif committed:
Scope autoScroll to LazyListState

ovitrif committed:
Put args of scrollToItem on separate lines

ovitrif committed:
Rearrange code in AutoScrollingLazyColumn

ovitrif committed:
Disable user scroll on large texts

ovitrif committed:
Move jetpack logo content description text to jetpack strings

ovitrif committed:
Add and use isOdd Int extension

ovitrif committed:
Override AutoScrollingLazyColumn divider for LargeTexts list

ovitrif committed:
Use textStyle to style autoscrolling large text

ovitrif committed:
Extract scroll offset

ovitrif committed:
Replace error in if statement with contract check

ovitrif committed:
Order compose margin units ascending

ovitrif committed:
Style buttons panel on Jetpack landing screen

ovitrif committed:
Extract ButtonsColumn composable

ovitrif committed:
Rename divider to itemDivider in AutoScrollingLazyColumn

ovitrif committed:
Move modifier argument in AutoScrollingLazyColumn

ovitrif committed:
Reformat code for ButtonsColumn

ovitrif committed:
Set large font size to 44 to match design text wrapping

ovitrif committed:
Add blur effect to splash background

ovitrif committed:
Wrap LandingScreen content in a Box to fix top fade gradient size

ovitrif committed:
Call ButtonsColumn content in composable body to cleanup code

ovitrif committed:
Set height of the blurred area equal to the height of the buttons container

ovitrif committed:
Fix blur area clipping

ovitrif committed:
Fix blur area clipping

ovitrif committed:
Fix large text font style

ovitrif committed:
Use a spacer of 2 dp between large texts

ovitrif committed:
Fix auto scroll

ovitrif committed:
Revert auto scrolling direction

ovitrif committed:
Add scrollBy and scrollDelay params to AutoScrollingLazyColumn

ovitrif committed:
Rename sub lists

ovitrif committed:
Extract string resources

ovitrif committed:
Set opacity of blurred text to .5

mkevins committed:
Deconstruct splash background box composable

This is necessary to change the structure a bit: the buttons composable
must be a sibling to the clipped background to utilize subcomposition.

mkevins committed:
Move sub components out of column and use box alignment instead

mkevins committed:
Use subcomposition to avoid recomposition for sibling measurement

This approach replaces the onSizeChanged approach to avoid recomposition
in order to measure the buttons composable's height for setting the
clipping height on the blurred background. Some restructuring was
necessary to make these composables direct siblings.

ovitrif committed:
Cleanup and optimize imports after refactor to SubcomposeLayout

ovitrif committed:
Extract reusable AutoScrollingLazyColumn component

ovitrif committed:
Apply transparency to the text behind buttons only when blur is not supported

ovitrif committed:
Refactor to extract FrostedGlassColumn component

ovitrif committed:
Don't blur behind frosted glass top border to give it more glassy look

ovitrif committed:
Move SlotsEnum to FrostedGlassColumn file

ovitrif committed:
Add missing check for getLoginPrologueRevampedFragment

ovitrif committed:
Rename FrostedGlassColumn to ColumnWithFrostedGlassBackground

ovitrif committed:
Remove redundant @Composable annotations for subcompose arguments

ovitrif committed:
Clear full screen flags when exiting the view

ovitrif committed:
Fine-tune the top linear gradient on the jp splash screen

ovitrif committed:
Fine-tune the top border of the frosted glass panel

ovitrif committed:
Replace forEach with items function in AutoScrollingListItem

ovitrif committed:
Remove single modifier param name in LoopingTextWithBackground box

ovitrif committed:
Extract private composable for rememberLargeTextItems

ovitrif committed:
Remove log statement

ovitrif committed:
Change scroll direction when scrollBy is negative

ovitrif committed:
Remove double quotes from large text resources

ovitrif committed:
Rename IntExtensions to NumberExtensions including FLoat.isNegative

ovitrif committed:
Refactor scrolling text to work in any direction

mkevins committed:
Add view model for emitting acceration sensor data

This will be useful for the revamped Jetpack landing screen's animation.

mkevins committed:
Implement repeating column composable

This will repeat the contents, and offset them vertically by a specified
amount. This will be useful for animating the background views in the
Jetpack landing screen.

mkevins committed:
Implement sensor driven animated composable for background text

This refactors the implementation to use a simplified repeating column
composable, offset by values derived by sensor-driven live data.

mkevins committed:
Add velocity decay and drift to animation

mkevins committed:
Lock LoginActivity to portrait mode

This is uesful to avoid orientation changes during the Jetpack
interactive landing screen animation, but also affects the WordPress
app. Note that this will be locked in portrait even when the feature
flag is not enabled. This should be ok, imo, since the UI in landscape
is not ideal anyway.

mkevins committed:
Extract physics model to view model and use CompositionLocal

This refactors the physics model to the view model for the fragment, and
also makes the consuming composables depend on position, rather than
acceleration. This avoids a duplication of state in the normal and
blurred composables, preventing the possibility that they can drift
"out-of-sync" with one another, since they previously each maintained
their own source of truth regarding the offset position.

This change also enables "donut hole skipping" while making use of
CompositionLocal for the position, allowing the intermediate composables
in the hierarchy to be completely decoupled from the position updates.

ovitrif committed:
Provide LoginPrologueRevampedViewModel with HILT

ovitrif committed:
Fix typo in ACCELERATION_FACTOR const

ovitrif committed:
Fix whitespaces in compose code

ovitrif committed:
Reformat JP landing screen composables code

ovitrif committed:
Extract scroll position updates and provider to composable

ovitrif committed:
Merge branch 'issue/17073-jetpack-landing-screen-revamp' into issue/17075-jetpack-landing-screen-sensor-animation

# Conflicts:
#	WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedFragment.kt
#	WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/components/ColumnWithFrostedGlassBackground.kt
#	WordPress/src/jetpack/java/org/wordpress/android/ui/compose/components/AutoScrollingLazyColumn.kt
#	WordPress/src/main/java/org/wordpress/android/ui/accounts/LoginActivity.java

ovitrif committed:
Post-merge fixes

ovitrif committed:
Provide frosted-glass background composable from outer scope

ovitrif committed:
Fix code formatting in LoginPrologueRevampedViewModel

ovitrif committed:
Merge branch 'feature/revamp-landing-screen' into issue/17075-jetpack-landing-screen-sensor-animation

# Conflicts:
#	WordPress/src/jetpack/java/org/wordpress/android/ui/compose/components/AutoScrollingLazyColumn.kt

ovitrif committed:
Use destructuring to better outline sensor data

ovitrif committed:
Add background argument name to ColumnWithFrostedGlassBackground call

ovitrif committed:
Make position private in LoginPrologueRevampedViewModel

ovitrif committed:
Clear formatting in _positionData

ovitrif committed:
Remove bg_jetpack_login_splash.png since it's replaced by density-specific image set

ovitrif committed:
Remove unused import in AppComponent

ovitrif committed:
Merge branch 'feature/revamp-landing-screen' into issue/17075-jetpack-landing-screen-sensor-animation

ovitrif committed:
Merge branch 'feature/revamp-landing-screen' into issue/17075-jetpack-landing-screen-sensor-animation

# Conflicts:
#	WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedFragment.kt
#	WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/components/ColumnWithFrostedGlassBackground.kt
#	WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/components/LoopingText.kt

ovitrif committed:
Merge branch 'feature/revamp-landing-screen' into issue/17075-jetpack-landing-screen-sensor-animation

# Conflicts:
#	WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/components/LoopingText.kt

ovitrif committed:
Merge branch 'feature/revamp-landing-screen' into issue/17075-jetpack-landing-screen-sensor-animation

ovitrif committed:
Make sure text scrolls when sensors are off

antonis reviewed (commented):
Great work @ovitrif 👍 

I started reviewing the changes and testing the app. I noticed that the text moves too fast on the initial position. This is more of a personal opinion since I may have not followed all the conversations on this but wanted to confirm that what I'm reproducing on Pixel 5 (Android 13) below is the expected behaviour or the speed may be different per device. In the video I started with my normal holding position (~ 60 degrees) and ended with the device flat on my desk. 

https://user-images.githubusercontent.com/304044/192317085-525f8f05-e74b-4a97-b199-e28fd2ce3662.mp4

I'll continue the review and get back to you.

ovitrif commented:
> I started reviewing the changes and testing the app. I noticed that the text moves too fast on the initial position. This is more of a personal opinion since I may have not followed all the conversations on this but wanted to confirm that what I'm reproducing on Pixel 5 (Android 13) below is the expected behaviour or the speed may be different per device. In the video I started with my normal holding position (~ 60 degrees) and ended with the device flat on my desk.

Your point is perfectly valid to me @antonis, I haven't tweaked the speed on top of @mkevins implementation with Accelerometer sensor which is the source of this PR — therefore the credit should go to Matthew for this great work 🚀 — but we were discussing this last week.

I wanted to get early feedback and proceed to make adjustments with a more informed stance, so this worked quite great, thank you for sharing your view 👍

I'll look into options to tweak the speed for the most expected / normal phone vertical orientation.

ovitrif committed:
Decrease maximum velocity

ovitrif commented:
> I'll look into options to tweak the speed for the most expected / normal phone vertical orientation.

I [decreased the maximum velocity](https://github.com/wordpress-mobile/WordPress-Android/pull/17207/commits/4e5059ff3cd8ed6d81505395c2d3cd189870b67b) — thus obtaining a slower speed @ normal vertical orientation
I also tested the scenario with sensors off and `null` (it required a small fix — a default value for the `acceleration`).

@mkevins Let me know if [this change](https://github.com/wordpress-mobile/WordPress-Android/pull/17207/commits/2817914dbf06b0a2a2cc00dac84050c1ecbd52b9) is aligned with the model 🙇 

P.S. From my side I think this PR is ready for review, the only reason it's a draft is because https://github.com/wordpress-mobile/WordPress-Android/pull/17203 should be merged first into `trunk`, and then this one should target `trunk` 🎉 .

P.P.S. Added Test Case 3️⃣ for testing the animation when sensors are disabled.

ovitrif committed:
Merge branch 'feature/revamp-landing-screen' into issue/17075-jetpack-landing-screen-sensor-animation

ovitrif committed:
Update release notes for PR 17207

ovitrif committed:
Merge branch 'feature/revamp-landing-screen' into issue/17075-jetpack-landing-screen-sensor-animation

ovitrif committed:
Update release notes for PR 17203

mkevins commented:
> @mkevins Let me know if [this change](https://github.com/wordpress-mobile/WordPress-Android/pull/17207/commits/2817914dbf06b0a2a2cc00dac84050c1ecbd52b9) is aligned with the model 🙇

Indeed this is looking good. Thanks for the fix for the non-sensor version. The reason it wasn't working well, after testing on an emulator without sensors is that the `DRIFT` was being applied in the `onSensorChanged` function instead of inside the `updateForFrame` function. This meant that the `DRIFT` was essentially absent from devices without sensor data on.

The following patch demonstrates a more appropriate place for the `DRIFT` to be applied:

<details>
<summary>Patch to apply DRIFT correctly</summary>

```patch
diff --git a/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt b/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
index 99e8570448..3e69a75d12 100644
--- a/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
+++ b/WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedViewModel.kt
@@ -31,7 +31,7 @@ private const val DRIFT = -0.05f
 class LoginPrologueRevampedViewModel @Inject constructor(
     @ApplicationContext appContext: Context,
 ) : ViewModel() {
-    private var acceleration = -0.3f // Default value when sensor is off
+    private var acceleration = 0f
     private var velocity = 0f
     private var position = 0f
 
@@ -47,7 +47,7 @@ class LoginPrologueRevampedViewModel @Inject constructor(
      */
     fun updateForFrame(elapsed: Float) {
         // Update the velocity, (decayed and clamped to the maximum)
-        velocity = (velocity * exp(elapsed * VELOCITY_DECAY) + elapsed * acceleration)
+        velocity = (velocity * exp(elapsed * VELOCITY_DECAY) + elapsed * (acceleration + DRIFT))
                 .coerceIn(-MAXIMUM_VELOCITY, MAXIMUM_VELOCITY)
         // Update the position, modulo 1 (ensuring a value greater or equal to 0, and less than 1)
         position = ((position + elapsed * velocity) % 1 + 1) % 1
@@ -73,7 +73,7 @@ class LoginPrologueRevampedViewModel @Inject constructor(
 
         override fun onSensorChanged(event: SensorEvent?) {
             event?.values?.let { (_, yAxisAcceleration, _) ->
-                acceleration = yAxisAcceleration * ACCELERATION_FACTOR + DRIFT
+                acceleration = yAxisAcceleration * ACCELERATION_FACTOR
                 postValue(position)
             }
         }
```
</details>

However, the `DRIFT` may have a slightly different purpose than the default / initial acceleration value. The `DRIFT` variable biases the acceleration such that a device  _with_ sensor data on will still result in scrolling text when the device is flat on a table. The value of this bias influences the behavior of the interactive animation in an ongoing way.

So, the way that you have addressed the "sensor-less" case has the advantage that we can configure different scrolling speeds for devices without sensors and for devices with sensors that happen to be flat on a table. For example, we may want that "flat on a table" case to be _slower_ than the default "sensor-less" case. I hope this makes sense. tl;dr, I think your change to the default acceleration is a great solution 😄

ovitrif committed:
Restore preview functionality by keeping layout stateless

ovitrif committed:
Move PositionProvider above LoginScreenRevamped

ovitrif commented on WordPress/src/jetpack/java/org/wordpress/android/ui/accounts/login/LoginPrologueRevampedFragment.kt:
```diff
@@ -64,6 +77,35 @@ class LoginPrologueRevampedFragment : Fragment() {
+        LaunchedEffect(Unit) {
+            var lastFrameNanos: Long? = null
+            while (isActive) {
+                val currentFrameNanos = awaitFrame()
```
@mkevins I've found this issue in the Android Public Tracker:
[Add a lint check for using awaitFrame inside Compose code 
](https://issuetracker.google.com/issues/229748038)

It makes an argument on why we shouldn't use `awaitFrame` (hint: I was researching `awaitFrame`) in Compose code, but rather the `withFrameNanos` from Compose's own time source: [MonotonicFrameClock](https://developer.android.com/reference/kotlin/androidx/compose/runtime/MonotonicFrameClock#withFrameNanos(kotlin.Function1)).

I will adapt the code to use that instead and test if it still works as expected

ovitrif replied:
Update: Still works nicely, I applied the change in https://github.com/wordpress-mobile/WordPress-Android/pull/17207/commits/20e21a4b3964054f2898e56e0833cdf53249a6b1

mkevins replied:
Oh nice find Ovi! This is a good improvement that will helpful for supporting possible future UI testing 👍

ovitrif committed:
Replace awaitFrame with withFrameMillis from MonotonicFrameClock

This public Android issue make an argument why we shouldn't use awaitFrame in Compose: https://issuetracker.google.com/issues/229748038

antonis reviewed (approved):
Great work @ovitrif 👍 
I've retested with the latest changes and everything worked great for me. The code also LGTM 🎉 

Note: I'm still able to reproduce the bars issue [but this seems fixed in the next design tweaks PR.](https://github.com/wordpress-mobile/WordPress-Android/pull/17231#issuecomment-1262207344)

ovitrif commented:
> Great work @ovitrif 👍 I've retested with the latest changes and everything worked great for me. The code also LGTM 🎉
> 
> Note: I'm still able to reproduce the bars issue [but this seems fixed in the next design tweaks PR.](https://github.com/wordpress-mobile/WordPress-Android/pull/17231#issuecomment-1262207344)

Thank you for the review @antonis 🙇 

I've created a separate ticket on the project board to test more intensively the issue with the system bars.

Let's give it another testing round after merging the remaining implementation work, when working on that task 👍
